themes/default/content/blog/converting-full-terraform-programs-to-pulumi/index.md

```diff
@@ -0,0 +1,524 @@
+
+# Set the Hostname
+hostnamectl set-hostname "{namePrefix}-$INSTANCE_ID"
+systemctl restart rsyslog.service
+
+# Get Configuration and SSL Certs
+aws --region {region} s3 cp s3://{vaultResourceId}/resources/config/config.hcl {vaultConfigDir}/config.hcl
+
+...
+""".format(**args)
+
+userdata = pulumi.Output.all(
+    namePrefix=args["namePrefix"],
+    region=args["region"],
+    vaultResourceId=vault_resources.id,
+    vaultConfigDir=args["vaultConfigDir"],
+).apply(format_userdata)
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+type userDataArgs struct {
+	NamePrefix      string
+	Region          string
+	VaultResourceId string
+	VaultConfigDir  string
+}
+
+func generateUserData(args userDataArgs) (string, error) {
+	tmpl, err := template.New("userdata").Parse(`#!/bin/bash
+
+# Get the Instance ID
+INSTANCE_ID=$(curl http://169.254.169.254/latest/meta-data/instance-id)
+
+# Set the Hostname
+hostnamectl set-hostname "{{ .NamePrefix }}-$INSTANCE_ID"
+systemctl restart rsyslog.service
+
+# Get Configuration and SSL Certs
+aws --region {{ .Region }} s3 cp s3://{{ .VaultResourceId }}/resources/config/config.hcl {{ .VaultConfigDir }}/config.hcl
+
+...`)
+	if err != nil {
+		return "", err
+	}
+	buf := bytes.Buffer{}
+	err = tmpl.Execute(&buf, args)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+userdata := pulumi.All(
+	args.NamePrefix,
+	args.Region,
+	vaultResources.ID().ToStringOutput(),
+	args.VaultConfigDir,
+).ApplyT(func(args []any) (string, error) {
+	return generateUserData(userDataArgs{
+		NamePrefix:      args[0].(string),
+		Region:          args[1].(string),
+		VaultResourceId: args[2].(string),
+		VaultConfigDir:  args[3].(string),
+	})
+}).(pulumi.StringOutput)
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var userdata = Output.Tuple(args.NamePrefix, args.Region, (Input<string>)vaultResources.Id, args.VaultConfigDir).Apply(t =>
+{
+   var (namePrefix, region, vaultResourcesId, vaultConfigDir) = t;
+   return $@"#!/bin/bash
+
+# Get the Instance ID
+INSTANCE_ID=$(curl http://169.254.169.254/latest/meta-data/instance-id)
+
+# Set the Hostname
+hostnamectl set-hostname ""{ namePrefix }-$INSTANCE_ID""
+systemctl restart rsyslog.service
+
+# Get Configuration and SSL Certs
+aws --region { region } s3 cp s3://{ vaultResourcesId }/resources/config/config.hcl { vaultConfigDir }/config.hcl
+
+...
+";
+});
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+### Wrapping Up
+
+After addressing these and some other tweaks so that the code compiles, we can now run the converted program with `pulumi up` to provision the Vault installation with Pulumi.
+
+The converter has saved us a ton of time, converting over 1,000 lines of Terraform to a modern Pulumi language, with only a small number of manual fix-ups required. From here, we can leverage our IDE and compiler to further refactor and improve the code, one of the many benefits of Pulumi!
+
```

Comment:
Is there anything we can/should say about migrating state over?  I think we ended up with not really having a great story here yet.  What's the best we could suggest?  And if nothing, should we reference that this is coming soon?

I suspect in practice, this will be one of the top questions we get.

Reply:
I think we should say we're working on it and it's coming very soon. Should probably be the next target feature to release, I don't think it will be a lot of work a lot of the groundwork is in place already.

Reply:
I added a section on this, mentioning what you can do today with some existing helper code we have for TypeScript and Go (linking to the reference docs on that; though, need to update those docs to mention Go), and teasing that built-in support is coming soon.